File: src/middleware/rate-limit.ts
import { createMiddleware } from 'hono/factory';
import { findApiKey } from '../services/config-store.js';

/**
 * Rate limit configuration from environment variables.
 */
const RATE_LIMIT_MAX = Number(process.env.RATE_LIMIT_MAX) || 60;
const RATE_LIMIT_WINDOW_MS = Number(process.env.RATE_LIMIT_WINDOW_MS) || 60_000; // 1 minute default

/**
 * In-memory store for rate limiting.
 * Maps API key to an array of request timestamps (sliding window).
 */
const requestStore = new Map<string, number[]>();

/**
 * Cleans up expired timestamps from the sliding window.
 */
function cleanupTimestamps(timestamps: number[], windowMs: number): number[] {
  const now = Date.now();
  const cutoff = now - windowMs;
  return timestamps.filter((ts) => ts > cutoff);
}

/**
 * Rate limiting middleware using sliding window algorithm.
 *
 * Limits requests by API Key (from Authorization: Bearer <key> header).
 * Default: 60 requests per minute per API key (configurable via env vars).
 *
 * Response headers:
 * - X-RateLimit-Limit: Maximum requests per window
 * - X-RateLimit-Remaining: Remaining requests in current window
 * - X-RateLimit-Reset: Unix timestamp when the window resets
 *
 * Returns 429 with OpenAI-compatible error format when limit exceeded.
 */
export const rateLimitMiddleware = createMiddleware<{
  Variables: { apiKey?: string };
}>(async (c, next) => {
  // Get API key from context (set by apiKeyAuthMiddleware)
  const apiKey = c.get('apiKey');

  // If no API key in context, skip rate limiting (should not happen after auth middleware)
  if (!apiKey) {
    return next();
  }

  // 读取当前 Key 的独立限流配置，未配置则使用全局默认
  const keyConfig = findApiKey(apiKey);
  const limitMax = keyConfig?.rateLimitMax ?? RATE_LIMIT_MAX;
  const limitWindowMs = keyConfig?.rateLimitWindowMs ?? RATE_LIMIT_WINDOW_MS;

  const now = Date.now();
  const windowStart = now - limitWindowMs;

  // Get or initialize timestamps for this API key
  let timestamps = requestStore.get(apiKey) || [];

  // Clean up expired timestamps
  timestamps = cleanupTimestamps(timestamps, limitWindowMs);

  // 清除不再活跃的 Key 条目，避免内存泄漏
  if (timestamps.length === 0 && requestStore.has(apiKey)) {
    requestStore.delete(apiKey);
  }

  // Calculate remaining requests
  const currentCount = timestamps.length;
  const remaining = Math.max(0, limitMax - currentCount);

  // Calculate reset time (oldest timestamp in window + window duration)
  const oldestTimestamp = timestamps.length > 0 ? Math.min(...timestamps) : now;
  const resetTime = oldestTimestamp + limitWindowMs;

  // Set rate limit headers
  c.header('X-RateLimit-Limit', String(limitMax));
  c.header('X-RateLimit-Remaining', String(remaining));
  c.header('X-RateLimit-Reset', String(Math.ceil(resetTime / 1000))); // Unix timestamp in seconds

  // Check if rate limit exceeded
  if (currentCount >= limitMax) {
    const retryAfterSeconds = Math.ceil((resetTime - now) / 1000);

    return c.json(
      {
        error: {
          message: `Rate limit exceeded. Please retry after ${retryAfterSeconds} seconds.`,
          type: 'rate_limit_error',
          code: 'rate_limit_exceeded',
        },
      },
      429,
    );
  }

  // Record this request
  timestamps.push(now);
  requestStore.set(apiKey, timestamps);

  await next();
});

/**
 * 定期清理不再活跃的 API Key 条目，避免长期运行后内存泄漏。
 * 每 10 分钟扫描一次，删除窗口期内无请求记录的条目。
 */
const CLEANUP_INTERVAL_MS = 10 * 60_000; // 10 minutes

setInterval(() => {
  const now = Date.now();
  const cutoff = now - RATE_LIMIT_WINDOW_MS;
  for (const [key, timestamps] of requestStore) {
    const active = timestamps.filter((ts) => ts > cutoff);
    if (active.length === 0) {
      requestStore.delete(key);
    } else {
      requestStore.set(key, active);
    }
  }
}, CLEANUP_INTERVAL_MS).unref();

File: app/middleware/rate_limit.py
"""Rate limiting middleware using sliding window."""

from __future__ import annotations

import time
from collections import defaultdict
from typing import TYPE_CHECKING

from fastapi import Request

from app.config import settings
from app.exceptions import RateLimitError

if TYPE_CHECKING:
    from app.dependencies import AppDependencies


class RateLimiter:
    """Encapsulated rate limiter state — no module-level globals.

    All mutable state is instance-level, making the limiter testable
    and safe in multi-instance scenarios.
    """

    def __init__(self) -> None:
        self._request_store: dict[str, list[float]] = defaultdict(list)

    def _cleanup_timestamps(self, timestamps: list[float], window_ms: float) -> list[float]:
        """Remove expired timestamps."""
        cutoff = time.time() * 1000 - window_ms
        return [ts for ts in timestamps if ts > cutoff]

    def _purge_stale_keys(self, window_ms: float) -> None:
        """Remove keys whose timestamps have all expired."""
        cutoff = time.time() * 1000 - window_ms
        stale = [k for k, ts in self._request_store.items() if not ts or max(ts) <= cutoff]
        for k in stale:
            del self._request_store[k]

    async def check(self, request: Request, api_key: str) -> None:
        """Rate limit check. Raises HTTPException(429) if limit exceeded."""
        # Determine limits for this key via DI
        deps: AppDependencies | None = getattr(request.app.state, "deps", None)
        config_store = deps.config_store if deps else None
        key_config = config_store.find_api_key(api_key) if config_store else None
        limit_max = key_config.rate_limit_max if (key_config and key_config.rate_limit_max) else settings.rate_limit_max
        limit_window_ms = key_config.rate_limit_window_ms if (key_config and key_config.rate_limit_window_ms) else settings.rate_limit_window_ms

        now_ms = time.time() * 1000
        timestamps = self._cleanup_timestamps(self._request_store.get(api_key, []), limit_window_ms)

        current_count = len(timestamps)
        remaining = max(0, limit_max - current_count)

        if current_count >= limit_max:
            oldest = min(timestamps) if timestamps else now_ms
            reset_time = oldest + limit_window_ms
            retry_after = max(1, int((reset_time - now_ms) / 1000))

            # Log rate limit event
            if deps and deps.log_collector:
                client_ip = request.client.host if request.client else "-"
                await deps.log_collector.emit(
                    "rate_limit_hit",
                    f"Rate limit hit from {client_ip}",
                    context={"path": request.url.path},
                    api_key_id=api_key[:8] + "..." if api_key else None,
                    client_ip=client_ip,
                )

            raise RateLimitError(
                f"Rate limit exceeded. Please retry after {retry_after} seconds."
            )

        # Record this request
        timestamps.append(now_ms)
        self._request_store[api_key] = timestamps

        # Periodic cleanup: purge stale keys that have no recent timestamps
        # to prevent unbounded memory growth from many distinct API keys.
        if len(self._request_store) > 1000:
            self._purge_stale_keys(limit_window_ms)


async def rate_limit_dependency(request: Request, api_key: str) -> None:
    """Rate limit check as a dependency. Call after api_key_auth_dependency.

    Uses the RateLimiter instance from AppDependencies so that state is
    shared consistently through the DI container rather than a module-level
    singleton.
    """
    deps: AppDependencies | None = getattr(request.app.state, "deps", None)
    limiter = deps.rate_limiter if deps else RateLimiter()
    await limiter.check(request, api_key)
